File: apps/desktop/src/main/port.ts
import net from 'node:net';

const DEFAULT_API_PORT = 18321;
const PORT_ENV_VAR = 'PIPLUS_DESKTOP_PORT';

export async function getFreePort(host = '127.0.0.1'): Promise<number> {
  return await new Promise((resolve, reject) => {
    const server = net.createServer();
    server.once('error', reject);
    server.listen(0, host, () => {
      const address = server.address();
      if (!address || typeof address === 'string') {
        server.close(() => reject(new Error('Failed to allocate free port')));
        return;
      }
      const { port } = address;
      server.close((error) => {
        if (error) reject(error);
        else resolve(port);
      });
    });
  });
}

/**
 * Check whether a specific port is available on the given host.
 * Returns true if the port can be bound, false if it is already in use.
 */
export async function isPortAvailable(host: string, port: number): Promise<boolean> {
  return await new Promise((resolve) => {
    const server = net.createServer();
    server.once('error', () => resolve(false));
    server.listen(port, host, () => {
      server.close(() => resolve(true));
    });
  });
}

/**
 * Get a port with the following priority:
 * 1. Environment variable PIPLUS_DESKTOP_PORT (if set and available)
 * 2. Built-in default (18321) if available
 * 3. Random free port as fallback
 *
 * Returns an object with the port and a flag indicating whether
 * the preferred (stable) port was used.
 */
export async function getPreferredPort(
  host = '127.0.0.1',
): Promise<{ port: number; preferred: boolean }> {
  // 1. Try env var (must be an exact integer string, e.g. "18321")
  const envPort = process.env[PORT_ENV_VAR];
  if (envPort) {
    const trimmed = envPort.trim();
    if (/^\d+$/.test(trimmed)) {
      const parsed = Number.parseInt(trimmed, 10);
      if (parsed > 0 && parsed <= 65535) {
        if (await isPortAvailable(host, parsed)) {
          console.log(`[desktop] using preferred port from ${PORT_ENV_VAR}=${parsed}`);
          return { port: parsed, preferred: true };
        }
        console.warn(
          `[desktop] ${PORT_ENV_VAR}=${parsed} unavailable; falling back`,
        );
      } else {
        console.warn(
          `[desktop] ${PORT_ENV_VAR}=${trimmed} out of range 1-65535; ignoring`,
        );
      }
    } else {
      console.warn(
        `[desktop] ${PORT_ENV_VAR}=${JSON.stringify(envPort)} is not an integer; ignoring`,
      );
    }
  }

  // 2. Try built-in default
  if (await isPortAvailable(host, DEFAULT_API_PORT)) {
    console.log(`[desktop] using preferred API port ${DEFAULT_API_PORT}`);
    return { port: DEFAULT_API_PORT, preferred: true };
  }

  // 3. Fallback to random port
  const fallback = await getFreePort(host);
  console.warn(
    `[desktop] preferred API port ${DEFAULT_API_PORT} unavailable; ` +
      `falling back to random port ${fallback}; ` +
      `localStorage origin will differ across restarts`,
  );
  return { port: fallback, preferred: false };
}

export { DEFAULT_API_PORT, PORT_ENV_VAR };

File: apps/desktop-tauri/src-tauri/src/port.rs
use std::net::TcpListener;

const DEFAULT_API_PORT: u16 = 18321;
const PORT_ENV_VAR: &str = "PIPLUS_DESKTOP_PORT";

/// Check if a specific port is available on the given host.
pub fn is_port_available(host: &str, port: u16) -> bool {
    TcpListener::bind(format!("{}:{}", host, port)).is_ok()
}

/// Get a random free port from the OS.
pub fn get_free_port(host: &str) -> std::io::Result<u16> {
    let listener = TcpListener::bind(format!("{}:0", host))?;
    Ok(listener.local_addr()?.port())
}

/// Get a port with the following priority:
/// 1. Environment variable PIPLUS_DESKTOP_PORT (if set and available)
/// 2. Built-in default (18321) if available
/// 3. Random free port as fallback
///
/// Returns (port, is_preferred) — mirrors Electron's `getPreferredPort()`.
pub fn get_preferred_port(host: &str) -> (u16, bool) {
    // 1. Try env var
    if let Ok(env_val) = std::env::var(PORT_ENV_VAR) {
        let trimmed = env_val.trim().to_string();
        if let Ok(port) = trimmed.parse::<u16>() {
            if port > 0 && is_port_available(host, port) {
                eprintln!(
                    "[desktop] using preferred port from {}={}",
                    PORT_ENV_VAR, port
                );
                return (port, true);
            }
            if port > 0 {
                eprintln!(
                    "[desktop] {}={} unavailable; falling back",
                    PORT_ENV_VAR, port
                );
            }
        } else {
            eprintln!(
                "[desktop] {}={:?} is not a valid port; ignoring",
                PORT_ENV_VAR, trimmed
            );
        }
    }

    // 2. Try built-in default
    if is_port_available(host, DEFAULT_API_PORT) {
        eprintln!("[desktop] using preferred API port {}", DEFAULT_API_PORT);
        return (DEFAULT_API_PORT, true);
    }

    // 3. Fallback to random port
    match get_free_port(host) {
        Ok(port) => {
            eprintln!(
                "[desktop] preferred API port {} unavailable; \
                 falling back to random port {}; \
                 localStorage origin will differ across restarts",
                DEFAULT_API_PORT, port
            );
            (port, false)
        }
        Err(e) => {
            eprintln!(
                "[desktop] failed to get free port: {}; using default {}",
                e, DEFAULT_API_PORT
            );
            (DEFAULT_API_PORT, true)
        }
    }
}